iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -66,28 +69,73 @@ cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceD
     {
         auto& entry = m_serviceDescriptions.back();
         entry.emplace(serviceDescription);
+        (*entry).*count = 1U;
         return cxx::success<>();
     }
 
     return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
 }
 
-void ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
+cxx::expected<ServiceRegistry::Error>
+ServiceRegistry::addPublisher(const capro::ServiceDescription& serviceDescription) noexcept
+{
+    return add(serviceDescription, &ServiceDescriptionEntry::publisherCount);
+}
+
+cxx::expected<ServiceRegistry::Error>
+ServiceRegistry::addServer(const capro::ServiceDescription& serviceDescription) noexcept
+{
+    return add(serviceDescription, &ServiceDescriptionEntry::serverCount);
+}
+
+void ServiceRegistry::removePublisher(const capro::ServiceDescription& serviceDescription) noexcept
```

Comment:
This is possible and the exact semantics for the discovery are not yet known. That is why for the discovery we essentially project the result onto publishers only, to be consistent with the previous functionality. I discussed it with @elBoberido.

I will discuss with @elBoberido  how to handle servers in the `ServiceDiscovery` later, but this PR is concerned with the `ServiceRegistry` only. There will be a separate PR for this once the semantics are clear.

Reply:
@mossmaurice this is mostly about adding internal API so that I can continue on the Request-Response feature. Currently it is not decided how to communicate this to the user. It is only tracked in the `ServiceRegistry` and `findService` would return only `ServiceDescriptions` for entries from `Publisher`. So we remain at status quo.

@MatthiasKillat had the idea to make this information part of `ServiceDescription`. Initially I was not too fond of this idea and MaKi also feared that it would be too invasive to change the `ServiceDescription` so close to the code freeze since it might result in many changes. After I took counsel with my pillow I had an idea how we could have it in the `ServiceDescription` and do not need to change much of our code. The additional benefit would be that the current `findService` API could remain as is and also be extended in a source compatible way to find only `Publisher` or `Server`.

Let's assume we add an additional `messagingPattern` method to the `ServiceDescription`
```cpp
enum class MessagingPattern {
ANY,
PUB_SUB,
REQ_RES
};
```
With the current `findService` API the `ServiceDiscovery` would assume that the user wants `MessagingPattern::ANY` and the user needs to check the returned `ServiceDescription` with the `MessagingPattern ServiceDescription::messagingPattern()` method. The `findService` API could be extended with a fourth parameter with a default value like `MessagingPattern messagingPattern = MessagingPattern::ANY`. This means that if there is a Publisher with `{"Foo", "Bar", "Baz"}` and a Server with `{"Foo", "Bar", "Baz"}` the user would get two results for `ANY` and one result for either `PUB_SUB` or `REQ_RES`.

To be backward compatible and enable the user to reuse a `ServiceDescription` for `Publisher` and `Server`, the `ServiceDescription` will by default be created with `MessagingPattern::ANY`
```cpp
auto sd = ServiceDescription{"Foo", "Bar", "Baz"};
auto publisher = Publisher{sd, {}};
auto server = Server{sd, {}};
```
This make it possible to create method/field combinations for ara::com where a method can be used to update a field.

I'm not sure if the user should be able to set the `MessagingPattern` of the `ServiceDescription` by herself and would therefore initially only allow it for the `ServiceDiscovery` which would do it when the result is returned.

Even if we do not want to proceed with the suggestion of MaKi and me, this PR does not prevent us from going a different route and adding a second counter is currently the least invasive way to add this to the `ServiceRegistry`. Another option would be to have a second vector but that wouldn't make things easier and just increase the memory usage.

The user side has not to be decided now and we can have a more detailed discussion in person, together with MaKi and Spitzi.

Reply:
@elBoberido I agree. But this should not affect the current pull request as it is not at odds with your proposal. I also want to store it as memory efficient as (reasonably, i.e. no compression) possible to keep the registry small (unless we gain something like search speed).

Reply:
@mossmaurice I think we need to define the terms first. Look at ROS2 since we are having here the same terminology:
* Topic: Publish/Subscribe communication: https://docs.ros.org/en/foxy/Tutorials/Topics/Understanding-ROS2-Topics.html
* Service: Request/Response communication: https://docs.ros.org/en/foxy/Tutorials/Services/Understanding-ROS2-Services.html

With this in mind the phone-book analogy does not fit since these are two different things.

Our service registry mixes both Topics/Services and we have to distinguish between them, either we do this by adding a `MessagingPattern` or we have a `TopicRegistry` and a `ServiceRegistry`. The naming is also not perfect in our case.


Just for fun, what do you think happens when we do not distinguish those things, then we will receive bug reports like:
Oh I found service XYZ but the subscriber/client cannot connect to it. To solve this impediment we then have to communicate to the user what type of service is behind XYZ?
For me the best approach would be by introducing `MessagingPattern`.

Reply:
@elfenpiff Good point! I had a look at the `rmw` API last week. There are several `findService`-like functions available:

* `rmw_get_topic_names_and_types()` => `ANY`
* `rmw_get_service_names_and_types_by_node()` => `REQ_RES`
* `rmw_get_publisher_names_and_types_by_node()` => `PUB_SUB`

There are also two methods to get more info like QoS:

* `rmw_get_publishers_info_by_topic()`
* `rmw_get_subscriptions_info_by_topic()`

My latest view on how things should work in `v3.0` is:

We differentiate between a user `ServiceDescription` and an internal representation of `ServiceDescription` enriched with more meta information. The information flow would be the following:

1. User creates a `ServiceDescription service`, which only describes the data
2. User creates a `popo::Publisher(service, publisherOptions)`
3. RouDi combines the `ServiceDescription`, QoS and the info what kind of producer was used (`popo::Publisher` or `popo::ServerPort`) into an internal representation and stores it in the `ServiceRegistry`
4. User does a `ServiceDiscovery::findService` and the internal presentation is converted back to `ServiceDescription`
5. To get the QoS info we could a similar method like in ROS 2 e.g. `ServiceDiscovery::getOptionsForProducer(ServiceDescription& service)`

What do you think?

I've implemented the possibilty to search for different producer kinds on #1134

cc @budrus @MatthiasKillat @elBoberido

Reply:
@mossmaurice @elfenpiff could we please move this discussion to an issue or a github discussion? I just got notified about this and had to search for the right thread to find about it and I only did this because I knew that it must be in a closed thread. I'm not sure all find this, even when mentioned. At lease I would not have looked at the closed threads if I didn't knew it would be there.

Reply:
Created the [discussion](https://github.com/eclipse-iceoryx/iceoryx/discussions/1149).